external-testing/README.md

```diff
@@ -0,0 +1,60 @@
+# External Testing
+
+Let's talk about testing process against external resources. What does that
+mean? Well, talking to GitHub API, COPR API, running `fedpkg`, cloning
+repositories...
+
+
+## Use cases
+
+* Interact with an external API (such as GitHub, COPR, Pagure, FAS, k8s)
+  * Authenticated vs. unauthenticated calls.
+* Clone a git repository.
+* Construct a git repository
+* Run a command to perform an action.
+  * Changes happening only locally.
+  * RO access to a remote service.
+  * R/W access to a remote service.
+* Combination of multiple use cases into a single test case.
+
+
+## Solutions
+
+* flexmock
+
+* pytest fixtures
+
+* requre
+
+* VCRpy
+
+
+## VCRpy vs. requre
+
+The section you are craving to read. It's opinionated and not objective.
+
+| A feature | VCRpy | requre |
+|-----------|-------|--------|
+| Record & replay for python requests | ✓ | ✓ |
+| Match on call stack | ✘ | ✓ |
+| Match on URL & request type | ✓ | ✘ |
+| Record & replay for arbitrary functions | ✘ | ✓ |
+| Filtering sensitive data out of recorded responses | ✓ [1] | ✓ [2] |
+| Easy to use | ✓ [3] | ✘ [4] |
+| We control the project | ✘ | ✓ |
```

Comment:
Can you please try the use-case with multiple calls to the same API? E.g. test for creating new issue:
- list all issues (1.)
- create issue
- list all issues (2.)

Same call but with a different response.

Reply:
Another use-case:
- Different versions of library we use (e.g. `rebase-helper`, `pygithub`, `gitlab-python`, `copr`, ...) have a different set of calls. Can vrcpy handle this?

Reply:
ad second comment, that should be handled here: https://github.com/packit-service/packit/pull/785

ad first: I'll create a test in ogr, let me know if you have ideas what to test

Reply:
Maybe try the `create_issue` as I wrote -- I am interested in a situation where you have exactly the same call with two different responses.

In other words, how the vcrpy differentiates the two requests? (In you example, they differ a bit.)

Reply:
I did with the whoami call - vcrpy uses matching on the request, so it's not just url, it's also headers, params and data.

Reply:
Yes, but what if we want exactly same request, like this test:
1. get all issues (returns 42 issue objects)
2. create a new issue
3. get all issues  (returns 43 issue objects) -- same URL, same headers, same params and same data...

Reply:
I can try.